src/providers/ipfs/kubo.py

```diff
@@ -0,0 +1,83 @@
+# pylint: disable=duplicate-code
+
+import logging
+from json import JSONDecodeError
+
+import requests
+
+from .cid import CID
+from .types import FetchError, IPFSProvider, PinError, UploadError
+
+logger = logging.getLogger(__name__)
+
+
+class Kubo(IPFSProvider):
+    """Client for [Kubo](https://github.com/ipfs/kubo) IPFS"""
+
+    PORT_GATEWAY = 8080
+    PORT_RPC = 5001
+
+    # @see https://docs.ipfs.tech/reference/kubo/rpc/#api-v0-add
+    RPC_UNIXFS_ADD_ARGS: dict[str, int | str] = {
+        "chunker": "size-262144",
+        "hash": "sha2-256",
+        "cid-version": 0,
+        "trickle": "false",
+        "raw-leaves": "false",
+    }
+
+    def __init__(self, host: str, *, timeout: int) -> None:
+        super().__init__()
+        self.host = host
+        self.timeout = timeout
+
+    def fetch(self, cid: CID) -> bytes:
+        url = f"{self.host}:{self.PORT_GATEWAY}/ipfs/{cid}"
+        try:
+            resp = requests.get(url, timeout=self.timeout)
+            resp.raise_for_status()
+        except requests.RequestException as ex:
+            logger.error({"msg": "Request has been failed", "error": str(ex)})
+            raise FetchError(cid) from ex
+        return resp.content
+
+    def _upload(self, content: bytes, name: str | None = None) -> str:
+        # @see https://docs.ipfs.tech/reference/kubo/rpc/#api-v0-add
+
+        url = f"{self.host}:{self.PORT_RPC}/api/v0/add"
+        name = name or "file"  # The name doesn't make any difference.
+
+        try:
+            resp = requests.post(url, files={name: content}, params=self.RPC_UNIXFS_ADD_ARGS, timeout=self.timeout)
+            resp.raise_for_status()
+        except requests.RequestException as ex:
+            logger.error({"msg": "Request has been failed", "error": str(ex)})
+            raise UploadError from ex
+
+        try:
+            return resp.json()["Hash"]
+        except JSONDecodeError as ex:
+            raise UploadError from ex
+        except KeyError as ex:
+            raise UploadError from ex
+
+    def pin(self, cid: CID) -> None:
+        # @see https://docs.ipfs.tech/reference/kubo/rpc/#api-v0-pin-add
+
+        url = f"{self.host}:{self.PORT_RPC}/api/v0/pin/add"
```

Comment:
Can we make DRY here.
Seems exception catching and json parsing are same with `_upload` func

Reply:
There are subtle differences that makes a one-fits-all function not a good one.